### PR TITLE
make AssetGraph serializable

### PR DESCRIPTION
**Pull request by jakemac53**

### CODE REVIEW

File: lib/src/asset_graph/exceptions.dart [hunk 1]
```diff
@@ -11,3 +11,14 @@ class DuplicateAssetNodeException implements Exception {
...
 }
+
+class AssetGraphVersionError implements Exception {
+  final int versionSeen;
+  final int currentVersion;
+
+  AssetGraphVersionError(this.versionSeen, this.currentVersion);
+
+  @override
+  String toString() => 'AssetGraphVersionError: saw version $versionSeen but '
```
Comment (kegluneq) [inline lib/src/asset_graph/exceptions.dart]: Is it possible to be more concrete? What component of the system saw (read? deserialized?) the incorrect version?

Comment (jakemac53) [inline lib/src/asset_graph/exceptions.dart]: Actually I need to rename this to `Exception` not `Error`. In practice I plan on catching this Exception and just starting the build from scratch in this case, so the user should never see the message.

File: lib/src/asset_graph/node.dart [hunk 2]
```diff
@@ -16,32 +15,56 @@ class AssetNode {
 
   AssetNode(this.id);
 
+  factory AssetNode.deserialize(List serializedNode) {
```
Comment (kegluneq) [inline lib/src/asset_graph/node.dart]: I'll put in a quick plug for protos here. They're not perfect but you get backwards-compatible serialization, composition, etc. for free and their implementations will improve without you doing anything. I think it's worth the up front cost.

Comment (jakemac53) [inline lib/src/asset_graph/node.dart]: Ya I plan on evaluating protos soon, see https://github.com/dart-lang/build/issues/41.

Comment (kegluneq) [inline lib/src/asset_graph/node.dart]: Cool - I have experience with getting them set up, so feel free to reach out to chat about them when the time comes.

**Comment by kegluneq** [comment]
LGTM

**Pull request merged**